[export] Add ir spec

angelayi opened the pull request:
Summary: Copied IR spec over from Executorch

Test Plan: _docs_

Differential Revision: D49829187

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+
+Export IR is an intermediate representation (IR) for compilers, which bears
+similarities to MLIR and TorchScript. It is specifically designed to express the
+semantics of PyTorch programs that are written in Python. Export IR primarily
```
```suggestion
semantics of PyTorch programs. Export IR primarily
```

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+
+As in the example above, notice that each line has this format::
+
+   %<name>:[...] = <op_name>[target=<target>](args = (%arg1, %arg2, arg3, arg4, …)), kwargs = {})
```
Is this still accurate? I think a while ago we were talking about supporting kwargs?

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+- **Functional:** We say a callable is “functional” if it satisfies all the
+  following requirements:
+
+  - Non-aliasing: The output tensors do not share data storage with each other
```
Aliasing is allowed now, just not mutations. Let's remove this bullet point.

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+^^^^^^^^^^^
+
+Placeholder represents an input to a graph. Its semantics are exactly the same as in FX.
+Placeholder nodes must be the first N nodes in the nodes list of a graph. N can be zero.
```
N can be zero?

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+get_attr
+^^^^^^^^
+
+``get_attr`` nodes represent reading a submodule from the encapsulating
```
And constant tensors for now?

angelayi replied:
No? Constant tensors are lifted

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+   * - NamedTuple
+     - Python namedtuple
+   * - Custom class
+     - Any custom class defined with ``_register_pytree_node``
```
dataclass is missing here

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,501 @@
+   * - Custom class
+     - Any custom class defined with ``_register_pytree_node``
+
+Memory Formats
```
Do we need to document this? memory format produced are up to the aten ops semantics I think.

@tugsbayasgalan @bdhirsh does aot_export enforce any particular memory format at all?

cc. @kimishpatel

angelayi replied:
Oh yeah, we can remove it since dim_order wont be ready by PTC

gmagogsfm commented on docs/source/export.ir_spec.rst:
```diff
@@ -0,0 +1,471 @@
+   * - NamedTuple
+     - Python namedtuple
+   * - Dataclass
+     - Must be registered through ``register_dataclass``
```
Consider change this (and other func/class/method name) to a reference in rst

angelayi committed:
[export] Add ir spec (#110394)

Summary:

Copied IR spec over from Executorch
bypass-github-pytorch-ci-checks

Test Plan: _docs_

Reviewed By: ydwu4

Differential Revision: D49829187
